src/cluster_legacy.c

```diff
@@ -5702,29 +5707,31 @@ void addShardReplyForClusterShards(client *c, list *nodes) {
         addReplyArrayLen(c, 0);
     }
 
+    list *nodes = clusterGetNodesInMyShard(n);
+
     addReplyBulkCString(c, "nodes");
     addReplyArrayLen(c, listLength(nodes));
     listIter li;
     listRewind(nodes, &li);
     for (listNode *ln = listNext(&li); ln != NULL; ln = listNext(&li)) {
-        clusterNode *n = listNodeValue(ln);
-        addNodeDetailsToShardReply(c, n);
+        n = listNodeValue(ln);
+        addNodeDetailsToShardReply(c, n, topology);
         clusterFreeNodesSlotsInfo(n);
     }
+    listRelease(nodes);
 }
 
 /* Add to the output buffer of the given client, an array of slot (start, end)
  * pair owned by the shard, also the primary and set of replica(s) along with
  * information about each node. */
-void clusterCommandShards(client *c) {
-    addReplyArrayLen(c, dictSize(server.cluster->shards));
+void clusterCommandShards(client *c, int topology) {
+    serverAssert(server.cluster->nummasters > 0);
```

Comment:
I do not think server.cluster->nummasters > 0 checking is must.  0 should work, how do you think?

Reply:
Thank you very much for reviewing the changes. Will address the suggestions once everyone finishes reviewing the PR.